Wheel hub assemblies and wheel adjustments

Hirmuolio opened the pull request:
#### Summary

```SUMMARY: Balance "Vehicle wheels require vehicle mounts"```

#### Purpose of change

The skill requirements for replacing a wheel is too high. (note: Actual remove and install, not the less known "change tire")

The requirements to install a wheel on new spot is too low.

Repairing wheel with welder makes no sense.

#### Describe the solution

TL;DR:
* Wheels are attached to hub assemblies. Hub assemblies require skills and welder.
* Removing/installing wheel requires no skill. Just the wrench.
* Wheels are repaired with glue and plastic.
* Steerable wheels are (mostly) no more. Instead the wheel mount is steerable and a normal wheel is attached.

Now the wheel mount is separate part. the hub assembly.
Doing anything to the hub assembly requires few skill levels. Installing requires welder while removing requires hacksaw. Repairing requires welder.

Wheel mounts come in three variants:
* Light wheel mount for light wheels like bike/motorcycle wheels. The mount is easily crafted with welder, power drill and two scrap metal pieces.
* Hub assembly for normal car wheels (17" and 24"). This also has an associated item that you get when you remove it from the car. Uncraftable but can be disassembled for some metal and bearings.
* Heavy hub assembly for large car wheels (32" armored). Higher skill requirements than the normal one.

The wheel is bolted on the hub assembly. In practice this just means that the wheel part can be installed only if there is already a hub assembly of appropriate size in the spot.

Additionally each wheel mount has steerable variant. It is identical except for the steerable flag on it. Wheels attached on these mounts will act as steerable.  (old steerable wheels still work)

You can remove and install wheels on hub assemblies without any skills.

Casters and tricycle wheels don't need mounts.

Wheel repairing has also been adjusted. Wheels are repaired with glue and plastic.
Repairing skill requirements have also been adjusted a bit. Mostly they are similar to before.  
* Light (motorcycle/bike) 2
* Car wheels (17") 3
* Wide car wheels (24") 4
* Armored wheel (32") 5

Wheel install/remove/repair times are now (used to be 1h install, 30m remove 25m repair), :
* For all wheels except armored Install/remove/repair 15 min
* Armored wheel has install/repair/remove 20 min

Additional infrastructure changes:
* Vehicle part repair times can now be given as a string like "20 m".
* Flag `"NEEDS_JACKING"` on the part makes it require jacking (used to be hardcoded for wheels).
* The "change wheel" option has been removed from the vehicle window as it became unnecessary.
* Saves with wheels without wheel hubs will get wheel hubs added on them automatically.


#### Describe alternatives you've considered


#### Additional context
The install/remove/repair values were not defined in the json before this. The code has some legacy values for them that were used.  
I was going to put the time/repair adjustments separately but since they were not defined in json I had to add some definition for them in json so I put them here.

You can still use welder to repair wheels in inventory. There is no way to define how items are repaired.

Drum rollers and train wheels are left unchanged.

The save conversion will leave steerable wheels on vehicles from old saves. This will not cause any problems and can be left as is.

Open for value change suggestions.

kevingranade commented:
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/i-need-mechanincs-4-to-swap-a-tire-aka-wheel-rework/21138/8

ZhilkinSerg commented on src/vehicle.cpp:
```diff
@@ -1225,6 +1263,23 @@ bool vehicle::can_unmount( const int p, std::string &reason ) const
     }
 
+    //Can't remove a wheel mount if there's still a wheel there
+    if( part_flag( p, "WHEEL_MOUNT_LIGHT" ) &&
```
Can you make it a pair of two strings and iterate through collection of such pairs?

Hirmuolio replied:
I am not sure what you mean by this.

AMurkin replied:
Something like this, but with wheel data:
https://github.com/CleverRaven/Cataclysm-DDA/blob/60e0ca933a12015a36349f4c426472c1c39da6c6/src/weather.cpp#L924-L940

Hirmuolio replied:
I did that for checking wether the part can be removed.

Something similar could be done for the check wether the part can be installed in `vehicle::can_mount`. But that one is a bit messier.

I also noticed that some parts have different install and remove checks. For example alternator installing checks for `"E_ALTERNATOR"` flag but the removing check checks for `"ENGINE"` flag (probably impossible to cause problems but still a bit bad). There are also few things that require a part when installing but do not check for it when uninstalling.

The list of flags that require other flags should probably be defined in somewhere else too. Especially if the same thing is used by install and uninstall check. Maybe in JSON. But I'd rather leave that to someone else.

Hirmuolio commented:
Getting close to being done. I managed to move steering from steerable wheels to steerable wheel mounts. In the end it was just a small one line edit.

Vehicle efficiency test is a problem since the new parts change the vehicle weights slightly.
The numbers in this test are a mystery to me. They are at the end of vehicle_test.cpp.
The first value is the starting weight of the vehicle. I can fix that.
Rest of numbers are not the same numbers that get used in the test. The numbers that do not match in the test result are not anywhere in the tests.

kevingranade commented:
There's a "test" just above the efficiency test that generates median values for you, whenever we *know* a test is going to shift these values it's ok to just update them, though we'll want to eyeball the results to make sure they aren't unexpectedly impactful.

Hirmuolio commented:
All checks green and I think everything works.
This pull is ready to be looked at.

Night-Pryanik commented on data/json/vehicleparts/wheel.json:
```diff
@@ -1,4 +1,109 @@
+        "using": [ [ "welding_standard", 10 ] ]
+      },
+      "removal": {
+        "skills": [ [ "mechanics", 4 ] ],
```
I think mechanics 4 is a far too high requirement for a simple sawing off a hub.

Night-Pryanik commented on data/json/vehicleparts/wheel.json:
```diff
@@ -1,4 +1,109 @@
+      },
+      "removal": {
+        "skills": [ [ "mechanics", 4 ] ],
+        "time": "45 m",
```
Too much, IMO. Maybe 15 minutes?

Night-Pryanik commented on data/json/vehicleparts/wheel.json:
```diff
@@ -1,4 +1,109 @@
+      },
+      "removal": {
+        "skills": [ [ "mechanics", 5 ] ],
+        "time": "60 m",
```
The same for skills and time requirements.

Night-Pryanik commented on data/json/vehicleparts/wheel.json:
```diff
@@ -1,4 +1,109 @@
+        "time": "45 m",
+        "qualities": [ { "id": "WRENCH", "level": 2 }, { "id": "SAW_M", "level": 2 } ]
+      },
+      "repair": { "skills": [ [ "mechanics", 4 ] ], "time": "15 m", "using": [ [ "welding_standard", 5 ] ] }
```
Welding a broken hub should not require mechanics 4. I'd say mechanics 2 is enough for repairing all types of hubs.

Night-Pryanik commented on data/json/vehicleparts/wheel.json:
```diff
@@ -7,15 +112,19 @@
     "contact_area": 80,
-    "flags": [ "WHEEL" ],
+    "requirements": {
+      "install": { "skills": [ [ "mechanics", 1 ] ], "time": "10 m", "qualities": [ { "id": "WRENCH", "level": 2 } ] },
```
Why installing a metal wheel requires less time than installing most other wheel types? The same for removal and repairing.

Hirmuolio replied:
Added those values before others and forgot to update.
Metal wheel probably should be moved to obsolete items. It is uncraftable and is not found anywhere in the game (unless some hardcoded c++ location has it).

Hirmuolio commented:
Moved the metal wheel to obsolete (it is not used anywhere in json, mods or c++).
Reduced remove/repair durations and skill requirements. Mechanics 1 to remove light hub, mechanics 3 to remove medium/heavy hub, 15/20 min to remove medium/heavy hub.
Added mechanics 0 as skill requirement for wheels so you get some xp from changing tires.

Night-Pryanik commented:
> Added mechanics 0 as skill requirement for wheels so you get some xp from changing tires.

![изображение](https://user-images.githubusercontent.com/11132525/63878748-48eca580-c9db-11e9-9341-e0eaca2197db.png)

Hirmuolio commented:
That failed test is not related to this pull I think.

ZhilkinSerg commented:
When loading one of savegames:

```
 DEBUG    : init_vehicles: 'Animal Control Truck' part 'wheel_wide_steerable'(43) can't be installed to 2,-1
 
 FUNCTION : static void vehicle_prototype::finalize()
 FILE     : src/veh_type.cpp
 LINE     : 1081
```
